src/carousel/carousel.ts

```diff
@@ -0,0 +1,113 @@
+import {Component, Input, Output, Control, OnDestroy, EventEmitter, NgFor} from 'angular2/angular2';
+
+@Component({
+  selector: 'ngb-carousel, [ngb-carousel]',
+  directives: [NgFor],
+  template: `
+    <div class="carousel slide">
+      <ol class="carousel-indicators">
+        <li *ng-for="#slide of slides" [class.active]="slide.isActive"></li>
+      </ol>
+      <div class="carousel-inner" role="listbox">
+        <ng-content></ng-content>
+      </div>
+      <a class="left carousel-control" role="button" (click)="prev()">
+        <span class="icon-prev" aria-hidden="true"></span>
+        <span class="sr-only">Previous</span>
+      </a>
+      <a class="right carousel-control" role="button" (click)="next()">
+        <span class="icon-next" aria-hidden="true"></span>
+        <span class="sr-only">Next</span>
+      </a>
+    </div>
+  `
+})
+export class NgbCarousel {
+  private slides: NgbSlide[] = [];
+  private activeSlide: NgbSlide;
+
+  addSlide(slide: NgbSlide): void {
+    this.slides.push(slide);
+    if (slide.isActive || !this.activeSlide) {
+      this.select(slide);
+    }
+  }
+
+  removeSlide(slide: NgbSlide): void {
+    const index = this.slides.indexOf(slide);
+    if (index !== -1) {
+      this.slides.splice(index, 1);
+    }
+  }
+
+  select(slide: NgbSlide): void {
+    slide.isActive = true;
+    this.activeSlide = slide;
+    this.hideInactiveSlides();
+  }
+
+  next(): void {
+    let idx = this.slides.indexOf(this.activeSlide);
+    if (idx === this.slides.length - 1) {
+      this.select(this.slides[0]);
+    } else {
+      this.select(this.slides[idx + 1]);
+    }
+  }
+
+  prev(): void {
+    let idx = this.slides.indexOf(this.activeSlide);
+    if (idx === 0) {
+      this.select(this.slides[this.slides.length - 1]);
+    } else {
+      this.select(this.slides[idx - 1]);
+    }
+  }
+
+  hideInactiveSlides(): void {
+    this.slides.forEach((slide) => {
+      if (slide !== this.activeSlide) {
+        slide.isActive = false;
+      }
+    });
+  }
+}
+
+@Component({
+  selector: 'ngb-slide, [ngb-slide]',
+  host: {'(activeChange)': 'activeChange.next(isActive)'},
+  template: `
+    <div class="carousel-item" [class.active]="isActive">
+      <ng-content></ng-content>
+    </div>
+  `
+})
+export class NgbSlide implements OnDestroy {
+  @Input() private active: boolean = false;
+  @Output() private activeChange: EventEmitter = new EventEmitter();
+  private _isActive: boolean = false;
+
+  constructor(private carousel: NgbCarousel) {
```

Comment:
This is my point: I don't think we need to add it: it is enough that parent queries for children and maintains an index of an active one. This would mean that we don't need any intermediate data structures.